### PR TITLE
[SYCL][Doc] Update SYCL --offload-arch option.

### PR DESCRIPTION
**Pull request by srividya-sundaram**
Adopt LLVM Clang's `--offload-arch` option for `SYCL` offloading to `Intel GPUs, Intel CPUs, NVidia and AMD GPUs.`

### CODE REVIEW

File: sycl/doc/design/OffloadDesign.md [hunk 1]
```diff
@@ -355,6 +401,33 @@ Additional options passed by the user via the
 `-Xsycl-target-backend=spir64_x86_64 <opts>` command will be processed by a new
 option to the wrapper, `--cpu-tool-arg=<arg>`
 
+Similar to SYCL offloading to Intel GPUs using `--offload-arch`, SYCL AOT for Intel CPUs
+can also leverage the `--offload-arch` option.
+The valid CPU device names accepted for `--offload-arch` are CPU names from ``clang -march``.
+These names are more verbose, and do not overlap with the Intel GPU names.
+These user input CPU names are mapped to the corresponding ``openl-aot -march`` option.
```
Comment (mdtoguchi) [inline sycl/doc/design/OffloadDesign.md:408]:
```suggestion
These user input CPU names are mapped to the corresponding ``opencl-aot -march`` option.
```

File: sycl/doc/design/OffloadDesign.md [hunk 2]
```diff
@@ -355,6 +401,33 @@ Additional options passed by the user via the
 `-Xsycl-target-backend=spir64_x86_64 <opts>` command will be processed by a new
 option to the wrapper, `--cpu-tool-arg=<arg>`
 
+Similar to SYCL offloading to Intel GPUs using `--offload-arch`, SYCL AOT for Intel CPUs
+can also leverage the `--offload-arch` option.
```
Comment (mdtoguchi) [inline sycl/doc/design/OffloadDesign.md:405]:
```suggestion
will also leverage the `--offload-arch` option.
```

File: sycl/doc/design/OffloadDesign.md [hunk 3]
```diff
@@ -296,6 +296,52 @@ resemble `--gpu-tool-arg=<arch> <arg>`.  This corresponds to the existing
 option syntax of `-fsycl-targets=intel_gpu_arch` where `arch` can be a fixed
 set of targets.
 
+``--offload-arch``
+^^^^^^^^^^^^^^^^^^
+For SYCL offloading to Intel GPUs, Intel CPUs, NVidia and AMD GPUs, specify the device architecture using --offload-arch option. For instance 
+  ``--offload-arch=sm_80`` to target an NVidia Tesla A100, 
+  ``--offload-arch=gfx90a`` to target an AMD Instinct MI250X, or 
+  ``--offload-arch=sm_80,gfx90a`` to target both.
```
Comment (mdtoguchi) [inline sycl/doc/design/OffloadDesign.md:304]: What is the expected behavior when an invalid value is passed?

File: sycl/doc/design/OffloadDesign.md [hunk 4]
```diff
@@ -296,6 +296,52 @@ resemble `--gpu-tool-arg=<arch> <arg>`.  This corresponds to the existing
 option syntax of `-fsycl-targets=intel_gpu_arch` where `arch` can be a fixed
 set of targets.
 
+``--offload-arch``
+^^^^^^^^^^^^^^^^^^
```
Comment (mdtoguchi) [inline sycl/doc/design/OffloadDesign.md:300]: Looking at the rendered output, the `^^^^` doesn't seem to translate.

File: sycl/doc/design/OffloadDesign.md (hunk 3 above)
Comment (srividya-sundaram) [reply sycl/doc/design/OffloadDesign.md:304]:
```
clang++ -fsycl --offload-new-driver --offload-arch=sri syclfile.cpp 
clang++: error: SYCL target is invalid: 'sri'
```

File: sycl/doc/design/OffloadDesign.md [hunk 5]
```diff
@@ -355,6 +403,33 @@ Additional options passed by the user via the
...
+| Intel(R) Advanced Vector Extensions 2 | core-avx2 | avx2              |
+| Intel(R) Advanced Vector Extensions | corei7-avx | avx                |
+| Intel(R) Streaming SIMD Extensions 4.2 | corei7 | sse4.2              |
+| Intel(R) microarchitecture code name Westmere | westmere | wsm        |
+| Intel(R) microarchitecture code name Sandy Bridge | sandybridge | snb |
+| Intel(R) microarchitecture code name Ivy Bridge | ivybridge | ivyb    |
+| Intel(R) microarchitecture code name Broadwell | broadwell | bdw      |
+| Intel(R) microarchitecture code name Coffee Lake | coffeelake | cfl   |
+| Intel(R) microarchitecture code name Alder Lake | alderlake | adl     |
+| Intel(R) microarchitecture code name Skylake (server) | skx | skx     |
```
Comment (maarquitos14) [inline sycl/doc/design/OffloadDesign.md:426]: Is there a reason for having `--offload-arch` accepted value for Skylake to be short and not as descriptive as the rest? Feels like we're not consistent with the rest.

**Review by maarquitos14 — COMMENTED** [review]
I see we only have lists with the accepted value for `--offload-arch` for Intel devices, should we do the same for third party devices?

File: sycl/doc/design/OffloadDesign.md (hunk 5 above)
Comment (srividya-sundaram) [reply sycl/doc/design/OffloadDesign.md:426]: For SYCL offloading to Intel CPUs and GPUs, there is a mapping between the accepted values for `--offload-arch` and what is passed to the target-specific LLVM compiler (OCLOC or opencl-aot).
For NVidia GPUs and AMD GPUs, there is no such mapping. I can maybe add a list of accepted values for `--offload-arch `NVidia and AMD GPUs.

**Comment by srividya-sundaram** [comment]
> I see we only have lists with the accepted value for `--offload-arch` for Intel devices, should we do the same for third party devices?

For SYCL offloading to Intel CPUs and GPUs, there is a mapping between the accepted values for --offload-arch and what is passed to the target-specific LLVM compiler (OCLOC or opencl-aot).
For NVidia GPUs and AMD GPUs, there is no such mapping. I can maybe add a list of accepted values for --offload-arch NVidia and AMD GPUs.

File: sycl/doc/design/OffloadDesign.md (hunk 5 above)
Comment (srividya-sundaram) [reply sycl/doc/design/OffloadDesign.md:426]: I just noticed that Skylake client is missing from this list.
The `--offload-arch` accepted value for Skylake client is `skylake`, so `skx` was chosen for Skylake server.

**Comment by maarquitos14** [comment]
> For NVidia GPUs and AMD GPUs, there is no such mapping. I can maybe add a list of accepted values for `--offload-arch` for NVidia and AMD GPUs.

I think that would be helpful for users.

**Review by maarquitos14 — APPROVED** [review]

**Comment by mdtoguchi** [comment]
@gmlueck, please take a look as you were part of the original conversation for these values.

**Pull request closed without merging**